commitTurn duplicates messages already ingested from the transcript projection in 1.0.0-beta.6

## Environment

- lossless-claw: `1.0.0-beta.6`
- tag commit: `4dfadc1f33f86794168dcc17ea7c5d648fc215b0`
- OpenClaw: `2026.8.1-beta.3`
- Host path: native Codex app-server context-engine lifecycle
- SQLite integrity after reproduction: `quick_check: ok`; zero foreign-key violations

## Summary

When the transcript projection has already ingested a logical turn's messages, the first `commitTurn(...)` call for that accepted turn inserts the same messages again before writing the advancement receipt.

The advancement key remains idempotent on an identical `commitTurn` retry, but it does not deduplicate against messages persisted before the first receipt. In the minimal reproduction, three transcript-ingested messages become six after one successful `commitTurn`.

This appears to violate the documented content-based deduplication fallback for messages without a `stable_event_key`.

## Expected behaviour

`commitTurn(...)` should atomically record the advancement receipt while recognising that the supplied logical messages are already present at the conversation frontier.

For the reproduction:

- message rows after transcript ingestion: `3`
- message rows after `commitTurn`: `3`
- advancement receipts added: `1`
- duplicate identity-hash groups: `0`

## Actual behaviour

- message rows after transcript ingestion: `3`
- message rows after `commitTurn`: `6`
- advancement receipts added: `1`
- duplicate identity-hash groups: `3`; each group contains two rows
- `commitTurn` result: `{ status: "committed" }`

The copied-database reproduction recorded duplicate sequence pairs `1/4`, `2/5`, and `3/6` with matching identity hashes.

## Minimal reproduction

Add the attached self-contained test as `test/commit-turn-transcript-interleaving.test.ts` and run:

```bash
npm exec vitest -- run test/commit-turn-transcript-interleaving.test.ts
```

The core sequence is:

```ts
const engine = createEngine();
const messages = [
  makeMessage({ role: "user", content: "question", timestamp: 10_000 }),
  makeMessage({ role: "tool", content: "tool result", timestamp: 11_000 }),
  makeMessage({ role: "assistant", content: "answer", timestamp: 12_000 }),
];

for (const message of messages) {
  await engine.ingest({ sessionId, sessionKey, message });
}

expect(await storedMessages()).toHaveLength(3);
await engine.commitTurn(buildCommitParams(messages));
expect(await storedMessages()).toHaveLength(3); // beta.6 receives 6
```

Observed failure:

```text
AssertionError: expected [...] to have a length of 3 but got 6
Expected: 3
Received: 6
```

A second reproduction against a SQLite-consistent copy of a migrated production-size database produced the same `3 -> 6` result and three duplicate identity-hash groups. The evidence shared with this report is sanitised and contains no production content.

## Existing coverage gap

The existing `test/engine-commit-turn.test.ts` suite passes `15/15` on the same checkout. It covers:

- first `commitTurn` persistence;
- advancement-key retry idempotency;
- payload collision detection;
- tool-result preservation;
- ignored/stateless sessions.

It does not cover this order of operations:

1. transcript/live projection ingests messages;
2. first `commitTurn` for the same accepted turn runs;
3. advancement receipt is created without duplicating the already-persisted frontier.

The new self-contained regression fails while the existing suite remains green.

## Relevant code path

In `src/engine.ts`, `afterTurn(...)` reconciles the visible transcript and uses `alignRuntimeBatchAgainstCoveredFrontier(...)` or `deduplicateAfterTurnBatch(...)` before ingestion.

`commitTurn(...)` instead loops over every supplied message and calls `ingestSingle(...)` directly before inserting the receipt. It checks only whether the advancement key already exists. Therefore, a first receipt does not protect against prior transcript ingestion of the same turn.

A global unique index on `identity_hash` may be too broad because legitimately repeated content can occur in separate turns. The likely fix needs to deduplicate the supplied commit range against the covered conversation frontier while preserving the transaction and advancement-key collision guarantees.

## Impact

On a fully capable native host, one accepted turn can persist duplicate user and tool rows. This can:

- inflate the LCM database;
- repeat content during later assembly or summarisation;
- distort token estimates and compaction thresholds;
- weaken the intended one-logical-turn/one-durable-advancement semantics.

The defect was observed with user and tool rows. Assistant rows carrying a stable response ID can be protected by the `stable_event_key` unique index, but rows without such a key remain exposed.

## Suggested acceptance test

A fix should make both attached reproductions pass and retain the current `15/15` commit-turn suite:

1. ingest user/tool/assistant messages before `commitTurn`;
2. call `commitTurn` once;
3. assert no additional semantic message rows;
4. assert one advancement receipt;
5. retry identical `commitTurn` and assert `{ status: "duplicate" }` with no row changes;
6. change the payload under the same advancement key and assert the collision still fails closed;
7. verify `quick_check: ok` and no foreign-key violations.

## Reproduction artifacts

Two sanitised failing regression tests and a sanitised JSON evidence packet are available on request. No production database or private state is included in this issue.

--- a/.changeset/deduplicate-durable-turn-frontier.md
+++ b/.changeset/deduplicate-durable-turn-frontier.md
@@ -1,0 +1,5 @@
+---
+"@martian-engineering/lossless-claw": patch
+---
+
+Deduplicate host-accepted turn messages against the covered transcript frontier before recording a durable advancement, preventing native runtimes from storing a second copy when transcript projection ingestion wins the race while preserving unflushed suffixes and ambiguous data.

--- a/src/engine.ts
+++ b/src/engine.ts
@@ -4157,10 +4157,10 @@ export class LcmContextEngine implements ContextEngine {
           }
 
           // The visible transcript can reach LCM before the host advances the
-          // accepted turn. Only align when that race is proven by the current
-          // admission's host-written transcript entry id; content equality
-          // alone cannot distinguish a replay from a legitimate later turn
-          // that repeats the same messages.
+          // accepted turn. Only enter covered-frontier alignment when the
+          // current admission's host-written entry id anchors a trusted row.
+          // The aligner validates exact or decorated content and fails open on
+          // mismatches, so repeated later turns remain preserved.
           const conversation = await this.conversationStore.getConversationForSession({
             sessionId,
             sessionKey,
@@ -4186,8 +4186,7 @@ export class LcmContextEngine implements ContextEngine {
             (admissionAnchorTrust.trustState === "verified" ||
               admissionAnchorTrust.trustState === "repaired") &&
             admissionAnchor.role === params.admission.role &&
-            admissionAnchor.role === admissionMessage.role &&
-            admissionAnchor.content === admissionMessage.content;
+            admissionAnchor.role === admissionMessage.role;
           // Keep alignment inside the same transaction as the receipt. Its
           // ambiguous-partial-overlap path fails open to preserve new data.
           const messagesToIngest = transcriptAlreadyProjected

--- a/src/engine.ts
+++ b/src/engine.ts
@@ -4156,8 +4156,49 @@ export class LcmContextEngine implements ContextEngine {
             return { status: "duplicate" as const };
           }
 
+          // The visible transcript can reach LCM before the host advances the
+          // accepted turn. Only align when that race is proven by the current
+          // admission's host-written transcript entry id; content equality
+          // alone cannot distinguish a replay from a legitimate later turn
+          // that repeats the same messages.
+          const conversation = await this.conversationStore.getConversationForSession({
+            sessionId,
+            sessionKey,
+          });
+          const admissionMessage = toStoredMessage(params.messages[0]!);
+          const admissionAnchor = conversation
+            ? await this.conversationStore.getTranscriptEntryAnchorCandidate(
+                conversation.conversationId,
+                params.admission.entryId,
+              )
+            : null;
+          const admissionAnchorTrust = admissionAnchor
+            ? await this.conversationStore.getMessageTranscriptAnchorTrust(
+                admissionAnchor.messageId,
+              )
+            : null;
+          const transcriptAlreadyProjected =
+            conversation !== null &&
+            admissionAnchor !== null &&
+            admissionAnchorTrust !== null &&
+            admissionAnchorTrust.conversationId === conversation.conversationId &&
+            admissionAnchorTrust.transcriptEntryId === params.admission.entryId &&
+            (admissionAnchorTrust.trustState === "verified" ||
+              admissionAnchorTrust.trustState === "repaired") &&
+            admissionAnchor.role === params.admission.role &&
+            admissionAnchor.role === admissionMessage.role &&
+            admissionAnchor.content === admissionMessage.content;
+          // Keep alignment inside the same transaction as the receipt. Its
+          // ambiguous-partial-overlap path fails open to preserve new data.
+          const messagesToIngest = transcriptAlreadyProjected
+            ? await this.batchDeduplicator.alignRuntimeBatchAgainstCoveredFrontier(
+                sessionId,
+                sessionKey,
+                params.messages,
+              )
+            : params.messages;
           let ingestedCount = 0;
-          for (const message of params.messages) {
+          for (const message of messagesToIngest) {
             const result = await this.ingestSingle({
               sessionId,
               sessionKey,

--- a/test/engine-commit-turn.test.ts
+++ b/test/engine-commit-turn.test.ts
@@ -19,6 +19,7 @@ import {
   seedBacklogContext,
   tempDirs,
 } from "./helpers.js";
+import { attachTranscriptEntryMeta } from "../src/transcript.js";
 
 afterEach(cleanupEngineTestState);
 
@@ -132,6 +133,206 @@ describe("LcmContextEngine commitTurn", () => {
       message_count: 2,
       hash_length: 64,
     });
+  });
+
+  it("records the receipt without duplicating a turn already ingested from the transcript", async () => {
+    const engine = createEngine();
+    const params = buildCommitTurnParams();
+    params.messages = [
+      makeMessage({ role: "user", content: "current question", timestamp: 2_000 }),
+      makeMessage({ role: "tool", content: "current tool result", timestamp: 2_500 }),
+      makeMessage({ role: "assistant", content: "current answer", timestamp: 3_000 }),
+    ];
+    params.terminal = {
+      ...params.terminal,
+      activeMessagePosition: params.admission.activeMessagePosition + params.messages.length - 1,
+      rawSeq: params.admission.rawSeq + params.messages.length - 1,
+    };
+
+    for (const [index, message] of params.messages.entries()) {
+      await engine.ingest({
+        sessionId: params.sessionId,
+        sessionKey: params.sessionKey,
+        message: attachTranscriptEntryMeta(
+          { ...message },
+          {
+            entryId: index === 0 ? params.admission.entryId : `transcript-entry-${index + 1}`,
+            parentId: null,
+            timestamp: null,
+          },
+        ),
+      });
+    }
+    expect(await readStoredMessages(engine)).toHaveLength(3);
+
+    await expect(engine.commitTurn(params)).resolves.toEqual({ status: "committed" });
+    await expect(engine.commitTurn(params)).resolves.toEqual({ status: "duplicate" });
+
+    expect((await readStoredMessages(engine)).map((message) => message.content)).toEqual([
+      "current question",
+      "current tool result",
+      "current answer",
+    ]);
+    expect(
+      getEngineDatabase(engine)
+        .prepare("SELECT count(*) AS count FROM turn_advancements")
+        .get(),
+    ).toEqual({ count: 1 });
+    expect(
+      getEngineDatabase(engine)
+        .prepare("SELECT message_count FROM turn_advancements WHERE advancement_key = ?")
+        .get(params.advancementKey),
+    ).toEqual({ message_count: 0 });
+  });
+
+  it("ingests only the unflushed suffix before recording the receipt", async () => {
+    const engine = createEngine();
+    const params = buildCommitTurnParams();
+    params.messages = [
+      makeMessage({ role: "user", content: "current question", timestamp: 2_000 }),
+      makeMessage({ role: "tool", content: "current tool result", timestamp: 2_500 }),
+      makeMessage({ role: "assistant", content: "current answer", timestamp: 3_000 }),
+    ];
+    params.terminal = {
+      ...params.terminal,
+      activeMessagePosition: params.admission.activeMessagePosition + params.messages.length - 1,
+      rawSeq: params.admission.rawSeq + params.messages.length - 1,
+    };
+
+    await engine.ingest({
+      sessionId: params.sessionId,
+      sessionKey: params.sessionKey,
+      message: attachTranscriptEntryMeta(
+        { ...params.messages[0]! },
+        {
+          entryId: params.admission.entryId,
+          parentId: null,
+          timestamp: null,
+        },
+      ),
+    });
+
+    await expect(engine.commitTurn(params)).resolves.toEqual({ status: "committed" });
+
+    expect((await readStoredMessages(engine)).map((message) => message.content)).toEqual([
+      "current question",
+      "current tool result",
+      "current answer",
+    ]);
+    expect(
+      getEngineDatabase(engine)
+        .prepare("SELECT message_count FROM turn_advancements WHERE advancement_key = ?")
+        .get(params.advancementKey),
+    ).toEqual({ message_count: 2 });
+  });
+
+  it("preserves the full accepted turn when its projected admission anchor is suspect", async () => {
+    const engine = createEngine();
+    const params = buildCommitTurnParams();
+    await engine.ingest({
+      sessionId: params.sessionId,
+      sessionKey: params.sessionKey,
+      message: attachTranscriptEntryMeta(
+        { ...params.messages[0]! },
+        {
+          entryId: params.admission.entryId,
+          parentId: null,
+          timestamp: null,
+        },
+      ),
+    });
+    const conversation = await engine
+      .getConversationStore()
+      .getConversationBySessionId(params.sessionId);
+    expect(conversation).not.toBeNull();
+    if (!conversation) throw new Error("conversation missing");
+    const candidate = await engine
+      .getConversationStore()
+      .getTranscriptEntryAnchorCandidate(conversation.conversationId, params.admission.entryId);
+    expect(candidate).not.toBeNull();
+    if (!candidate) throw new Error("admission anchor missing");
+    await engine.getConversationStore().upsertMessageTranscriptAnchorTrust({
+      messageId: candidate.messageId,
+      conversationId: conversation.conversationId,
+      transcriptEntryId: params.admission.entryId,
+      trustState: "suspect",
+      source: "test",
+      reason: "legacy anchor has not been verified",
+    });
+
+    await expect(engine.commitTurn(params)).resolves.toEqual({ status: "committed" });
+
+    expect((await readStoredMessages(engine)).map((message) => message.content)).toEqual([
+      "current question",
+      "current question",
+      "current answer",
+    ]);
+    expect(
+      getEngineDatabase(engine)
+        .prepare("SELECT message_count FROM turn_advancements WHERE advancement_key = ?")
+        .get(params.advancementKey),
+    ).toEqual({ message_count: 2 });
+  });
+
+  it("preserves the full accepted turn when a trusted admission anchor has different content", async () => {
+    const engine = createEngine();
+    const params = buildCommitTurnParams();
+    await engine.ingest({
+      sessionId: params.sessionId,
+      sessionKey: params.sessionKey,
+      message: attachTranscriptEntryMeta(
+        makeMessage({ role: "user", content: "stale projected question", timestamp: 2_000 }),
+        {
+          entryId: params.admission.entryId,
+          parentId: null,
+          timestamp: null,
+        },
+      ),
+    });
+
+    await expect(engine.commitTurn(params)).resolves.toEqual({ status: "committed" });
+
+    expect((await readStoredMessages(engine)).map((message) => message.content)).toEqual([
+      "stale projected question",
+      "current question",
+      "current answer",
+    ]);
+    expect(
+      getEngineDatabase(engine)
+        .prepare("SELECT message_count FROM turn_advancements WHERE advancement_key = ?")
+        .get(params.advancementKey),
+    ).toEqual({ message_count: 2 });
+  });
+
+  it("preserves a later accepted turn that repeats identical content", async () => {
+    const engine = createEngine();
+    const first = buildCommitTurnParams({ advancementKey: "logical-turn-1" });
+    const second = buildCommitTurnParams({ advancementKey: "logical-turn-2" });
+    second.admission = {
+      ...second.admission,
+      entryId: "user-entry-2",
+      logicalTurnId: second.advancementKey,
+    };
+    second.terminal = {
+      ...second.terminal,
+      entryId: "assistant-entry-2",
+      effectiveParentId: second.admission.entryId,
+    };
+
+    await expect(engine.commitTurn(first)).resolves.toEqual({ status: "committed" });
+    await expect(engine.commitTurn(second)).resolves.toEqual({ status: "committed" });
+
+    expect((await readStoredMessages(engine)).map((message) => message.content)).toEqual([
+      "current question",
+      "current answer",
+      "current question",
+      "current answer",
+    ]);
+    expect(
+      getEngineDatabase(engine)
+        .prepare("SELECT count(*) AS count FROM turn_advancements")
+        .get(),
+    ).toEqual({ count: 2 });
   });
 
   it("preserves recurrent model-authored tool results across committed turns", async () => {

--- a/test/engine-commit-turn.test.ts
+++ b/test/engine-commit-turn.test.ts
@@ -185,6 +185,40 @@ describe("LcmContextEngine commitTurn", () => {
     ).toEqual({ message_count: 0 });
   });
 
+  it("aligns a decorated accepted user message with its bare projected admission", async () => {
+    const engine = createEngine();
+    const params = buildCommitTurnParams();
+    await engine.ingest({
+      sessionId: params.sessionId,
+      sessionKey: params.sessionKey,
+      message: attachTranscriptEntryMeta(
+        { ...params.messages[0]! },
+        {
+          entryId: params.admission.entryId,
+          parentId: null,
+          timestamp: null,
+        },
+      ),
+    });
+    params.messages[0] = makeMessage({
+      role: "user",
+      content: `[Sun 2026-01-01 12:00 GMT+0]\ncurrent question`,
+      timestamp: 2_000,
+    });
+
+    await expect(engine.commitTurn(params)).resolves.toEqual({ status: "committed" });
+
+    expect((await readStoredMessages(engine)).map((message) => message.content)).toEqual([
+      "current question",
+      "current answer",
+    ]);
+    expect(
+      getEngineDatabase(engine)
+        .prepare("SELECT message_count FROM turn_advancements WHERE advancement_key = ?")
+        .get(params.advancementKey),
+    ).toEqual({ message_count: 1 });
+  });
+
   it("ingests only the unflushed suffix before recording the receipt", async () => {
     const engine = createEngine();
     const params = buildCommitTurnParams();

